File: api/fetch.py
from bs4 import BeautifulSoup
import requests

def handler(request):
    if request.method != "POST":
        return {
            "statusCode": 405,
            "headers": {"Content-Type": "application/json"},
            "body": {"error": "Method Not Allowed. Use POST with JSON body"}
        }

    try:
        data = request.json()
        url = data.get("url")

        if not url:
            return {
                "statusCode": 400,
                "headers": {"Content-Type": "application/json"},
                "body": {"error": "Missing 'url' in request body"}
            }

        headers = {
            "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36"
        }

        response = requests.get(url, headers=headers, timeout=10)

        if response.status_code != 200:
            return {
                "statusCode": response.status_code,
                "headers": {"Content-Type": "application/json"},
                "body": {"error": f"Failed to fetch content. Status code: {response.status_code}"}
            }

        soup = BeautifulSoup(response.text, "html.parser")
        return {
            "statusCode": 200,
            "headers": {"Content-Type": "application/json"},
            "body": {"text": soup.get_text()}
        }

    except Exception as e:
        return {
            "statusCode": 500,
            "headers": {"Content-Type": "application/json"},
            "body": {"error": str(e)}
        }

File: api/fetch.js
const axios = require("axios");
const cheerio = require("cheerio");

module.exports = async (req, res) => {
  if (req.method !== "POST") {
    return res.status(405).json({ error: "Method not allowed. Use POST." });
  }

  const { url } = req.body;

  if (!url) {
    return res.status(400).json({ error: "Missing 'url' in request body." });
  }

  try {
    // Fetch full HTML
    const response = await axios.get(url, {
      headers: {
        "User-Agent":
          "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36"
      },
      responseType: "text",
      timeout: 10000
    });

    // Load HTML into Cheerio
    const html = response.data;
    const $ = cheerio.load(html);

    // Extract readable text
    const text = $("body").text().replace(/\s+/g, " ").trim();

    // Truncate if needed (for payload safety)
    const maxLength = 200000;
    const safeHTML = html.length > maxLength ? html.slice(0, maxLength) : html;
    const safeText = text.length > maxLength ? text.slice(0, maxLength) : text;

    // ✅ Return both HTML and parsed text
    return res.status(200).json({
      url,
      html: safeHTML,
      text: safeText,
      length: {
        html: safeHTML.length,
        text: safeText.length
      }
    });
  } catch (error) {
    return res.status(500).json({
      error: error.message || "Failed to fetch content."
    });
  }
};
